tools/add_triage_signature.py

```diff
@@ -1120,6 +1120,96 @@ def _update_ticket(self, url, issue_key, should_update=False):
         self._update_triaging_ticket(issue_key, report, should_update=should_update)
 
 
+class OSInstallationTime(Signature):
+    writing_image_start_event_regex = re.compile(r'.*reached installation stage Writing image to disk$')
+    writing_image_end_event_regex = re.compile(r'.*reached installation stage Writing image to disk: 100%$')
+    slow_threshold_seconds = 300
+
+    class NoEventFound(Exception):
+        pass
+
+    def __init__(self, jira_client):
+        super().__init__(jira_client, comment_identifying_string="h1. OS Installation Time:")
+
+    @classmethod
+    def _get_last_event(cls, all_events, regex):
+        """
+        We want to get the last event because sometimes the list of all events
+        for a triage ticket contains events from previous installations. The
+        last ones however, are always relevant to the ticket in question.
+        """
+        try:
+            *_, last = (
+                event
+                for event in all_events
+                if regex.match(event["message"]) is not None
+            )
+        except ValueError:
+            raise cls.NoEventFound
+
+        return last
+
+    @classmethod
+    def _get_last_event_timestamp(cls, all_events, regex):
+        return dateutil.parser.isoparse(cls._get_last_event(all_events, regex)["event_time"])
+
+    @classmethod
+    def _get_start_event_timestamp(cls, all_events):
+        return cls._get_last_event_timestamp(all_events, cls.writing_image_start_event_regex)
+
+    @classmethod
+    def _get_end_event_timestamp(cls, all_events):
+        return cls._get_last_event_timestamp(all_events, cls.writing_image_end_event_regex)
+
+    def _update_ticket(self, url, issue_key, should_update=False):
+        url = self._logs_url_to_api(url)
+        md = get_metadata_json(url)
+        cluster = md['cluster']
+        cluster_id = cluster['id']
+
+        events_by_host = defaultdict(list)
+        for event in get_events_json(cluster_url=url, cluster_id=cluster_id):
+            if "host_id" not in event:
+                # Cluster-level event, no host_id associated with it
+                continue
+            events_by_host[event["host_id"]].append(event)
+
+        hosts = []
+        for host in cluster['hosts']:
+            host_events = events_by_host[host["id"]]
```

Comment:
I wonder if we could have some generic function `get_events_by_host()` that would be accessible from different classes? That way if any other signature wants to parse events if will not need to have this logic of for-loop over all the events and building the map